streaming/consumer.cc

```diff
@@ -65,6 +65,12 @@ std::function<future<> (mutation_reader)> make_streaming_consumer(sstring origin
                         cf->enable_off_strategy_trigger();
                     }
                     return cf->add_sstable_and_update_cache(sst, offstrategy);
+                }).then([cf, sst] {
+                    // Streaming writer waits on split of newly added sstable if underlying table is in split mode.
+                    // FIXME: it's not safe to split after sstable was added, as a failure to split means split execute can happen later if
+                    //  the process that originated it give up due to the error. I think we need to wire the mutation splitter (consumer) instead,
+                    //  such that incoming data is split on the fly, and split sstables are added if in split mode.
+                    return cf->maybe_split_compaction_group_of(sst);
```

Comment:
Yes, it's unsafe to rely on this call here. Can we hide this in add_sstable_and_update_cache(), which will add sstables only after they are split?

Splitting on the fly is good, but it's just an optimization. Preparing could start after we already started writing an unsplit sstable, and we need to handle attaching of this unsplit sstable.

Reply:
I guess we should always check if the sstable to be added needs split whenever a sstable is added. Such check is cheap since we can check the [min_token, max_token] range of the sstable.

Reply:
I am worried that without 'splitting on the fly', a restart of this node could cause repair failure, the sstable would be loaded into the set, and split execute can then happen and find that sstable unsplit. But I think that can be definitely handled by doing it inside add_sstable_and_update_cache() during boot. So a SSTable is only added if it satisfies the split requirement. Will proceed with the change.

Reply:
> I guess we should always check if the sstable to be added needs split whenever a sstable is added. Such check is cheap since we can check the [min_token, max_token] range of the sstable.

Yeah, that's correct.